plugins/pipette/src/pipette.cc

```diff
@@ -0,0 +1,223 @@
+
+Pipette::Pipette()
+{
+}
+Pipette::~Pipette()
+{
+}
+
+Pipette::Ptr Pipette::create()
+{
+  return Ptr(new Pipette());
+}
+
+////////////////////////////////////////////////////////////////////////
+// PluginInformationInterface
+////////////////////////////////////////////////////////////////////////
+
+std::string Pipette::getPluginName()
+{
+  return "Pipette";
+}
+
+std::string Pipette::getPluginVersion()
+{
+  return "0.0";
+}
+
+void Pipette::registerCapabilities(ScroomPluginInterface::Ptr host)
+{
+  host->registerViewObserver("Pipette", shared_from_this<Pipette>());
+}
+
+////////////////////////////////////////////////////////////////////////
+// ViewObserver
+////////////////////////////////////////////////////////////////////////
+
+Scroom::Bookkeeping::Token Pipette::viewAdded(ViewInterface::Ptr view)
+{
+  PipetteHandler::Ptr handler = PipetteHandler::create();
+  view->registerSelectionListener(handler);
+  view->registerPostRenderer(handler);
+
+  view->addToolButton(GTK_TOGGLE_BUTTON(gtk_toggle_button_new_with_label("Pipette")), handler);
+
+  return Scroom::Bookkeeping::Token();
+}
+
+////////////////////////////////////////////////////////////////////////
+// PipetteHandler
+////////////////////////////////////////////////////////////////////////
+
+PipetteHandler::PipetteHandler()
+{
+  selection = nullptr;
+  enabled = false;
+  pendingJobs = ThreadPool::Queue::createAsync();
+}
+PipetteHandler::~PipetteHandler()
+{
+}
+
+PipetteHandler::Ptr PipetteHandler::create()
+{
+  return Ptr(new PipetteHandler());
+}
+
+void PipetteHandler::computeValues(ViewInterface::Ptr view)
+{
+  gdk_threads_enter();
+  view->setStatusMessage("Computing colors values...");
+  gdk_threads_leave();
+
+  // Get the image rectangle
+  auto presentation = boost::static_pointer_cast<PresentationInterface>(view->getCurrentPresentation());
+  if(presentation == nullptr)
+  {
+    printf("PANIC: Current presentation does not implement PresentationInterface!\n");
+    gdk_threads_enter();
+    view->setStatusMessage("Error when requesting the image data.");
+    gdk_threads_leave();
+    enabled = true;
+    return;
+  }
+  auto image = presentation->getRect().toIntRectangle();
+
+  // Get the selection rectangle
+  int sel_x = std::min(selection->start.x, selection->end.x);
+  int sel_y = std::min(selection->start.y, selection->end.y);
+  auto sel_rect = Scroom::Utils::Rectangle<int>(sel_x, sel_y, selection->width(), selection->height());
+
+  // Intersect both rectangles to get the part of the selection that overlaps the image
+  auto rect = sel_rect.intersection(image);
+
+  // Get the average color within the rectangle
+  auto pipette = boost::dynamic_pointer_cast<PipetteViewInterface>(presentation);
+  if(pipette == nullptr)
+  {
+    printf("PANIC: Presentation does not implement PipetteViewInterface!\n");
+    gdk_threads_enter();
+    view->setStatusMessage("Error when requesting the image data.");
+    gdk_threads_leave();
+    enabled = true;
```

Comment:
Why modify `enabled` here?

And... Wasn't `enabled` already `true` ?

Reply:
Enabled is used to ignore incoming events, given that we only want to handle one job at the same time `enabled` is set to `false` while a background job is running. If the if statement here fails then we want to make sure that we don't lock up the plugin before returning.